Capture pid in ProcessExitedException

GregPlowman opened the pull request:
Would it be possible to capture the worker pid in `ProcessExitedException`?
For my purpose, I would like to know the pid of a failed worker to ultimately determine the host machine.

This RFC is really asking whether this would be feasible and/or desirable.
Here's a first stab at it. I don't know how to test locally, so here it is raw.

GregPlowman committed:
Add pid to ProcessExitedException

GregPlowman committed:
Remove duplicate entry (same as line 79)

fredrikekre commented:
#27900 seems to do this.

GregPlowman commented:
Oh I see.

That PR is mostly concerned with command/pipeline exceptions, whereas this PR is about worker exceptions, although it seems the two have been combined over there.
Further discussion in the other PR.

GregPlowman commented:
Now that #27900 has taken the direction to untangle spawned process exceptions from Distributed worker exceptions, is it possible to review this PR and move forward?

GregPlowman commented:
With Julia 1.2 cut-off coming up, I'm not sure if it's a good or bad time to prod this.
Can someone review?

vtjnash commented on stdlib/Distributed/src/cluster.jl:
```diff
@@ -990,20 +990,20 @@ function _rmprocs(pids, waitfor)
-
 """
-    ProcessExitedException()
+    ProcessExitedException(pid::Int)
```
let's call this `worker_id`, this sounds too much to me like it would be the `getpid` value

vtjnash commented on stdlib/Distributed/src/cluster.jl:
```diff
@@ -990,20 +990,20 @@ function _rmprocs(pids, waitfor)
 After a client Julia process has exited, further attempts to reference the dead child will
 throw this exception.
 """
-ProcessExitedException()
```
Is this a stable API? We may need to define `ProcessExitedException() = ProcessExitedException(-1)`

GregPlowman replied:
It seems this call to `ProcessExitedException()` was here to anchor the docstring above it.
Wouldn't it have been creating an object in global scope that was never used?
In any case, I removed it and placed the docstring before the struct definition.

When would you create a `ProcessExitedException` without the worker id? If there's no good reason for it, I think it would be better to not provide a no-args constructor.

fredrikekre replied:
The point is that `ProcessExitedException()` works on Julia 1.0, and 1.1 but will throw an error after this PR.

GregPlowman replied:
Oh right, for backwards compatibility for code outside of Base.
No-arg constructor `ProcessExitedException() = ProcessExitedException(-1)` added.

GregPlowman committed:
ProcessExitedException - rename pid to worker_id

GregPlowman commented:
Rebased and CI all green (except Travis which is known to be broken)

GregPlowman committed:
Add no-arg constructor for backwards compatibility

StefanKarpinski commented:
The macOS Travis failure looks related:
```
┌ Warning: rmprocs: process 1 not removed
└ @ Distributed ~/build/JuliaLang/julia/usr/share/julia/stdlib/v1.2/Distributed/src/cluster.jl:1005
      From worker 42:	MethodError: Cannot `convert` an object of type Distributed.RRID to an object of type Int64
      From worker 42:	Closest candidates are:
      From worker 42:	  convert(::Type{T<:Number}, !Matched::T<:Number) where T<:Number at number.jl:6
      From worker 42:	  convert(::Type{T<:Number}, !Matched::Number) where T<:Number at number.jl:7
      From worker 42:	  convert(::Type{T<:Integer}, !Matched::Ptr) where T<:Integer at pointer.jl:23
      From worker 46:	MethodError: Cannot `convert` an object of type Distributed.RRID to an object of type Int64
```

GregPlowman committed:
Correct arg to ProcessExitedException

GregPlowman commented:
> The macOS Travis failure looks related:

Thanks for catching this. I think this is now fixed.

Rebased again, but still error on `macos64`. Can someone help interpret this error?

StefanKarpinski commented:
@staticfloat: can you figure out what's going on with the macOS tester? I think we're probably ok since Travis macOS CI passed.

staticfloat commented:
It's a problem with one of our macminis (the `-macmini` worker).  I'm addressing it, but in the meantime, I restarted the job, and it's using a different worker (the `-akatsuki` worker) so it should turn green in an hour or so.

GregPlowman commented:
All green, amazing!
Thanks Stefan and Elliot.
Ready to merge from my end.
